File: mcp-server/src/services/checker.ts
import { analyzeWork } from "./analyze.js";
import { buildFetcher, CrossrefError, DoiNotFoundError, type Fetcher } from "./crossref.js";
import type { RetractionStatus } from "./models.js";

/** Check a single DOI. Never throws for an expected failure -- becomes a NOT_FOUND/ERROR verdict instead. */
export async function checkDoi(doi: string, fetch?: Fetcher): Promise<RetractionStatus> {
  const fetcher = fetch ?? buildFetcher();
  try {
    const message = await fetcher(doi);
    return analyzeWork(message);
  } catch (err) {
    if (err instanceof DoiNotFoundError) {
      return { doi, verdict: "not_found", title: null, signals: [], error: err.message };
    }
    if (err instanceof CrossrefError) {
      return { doi, verdict: "error", title: null, signals: [], error: err.message };
    }
    throw err;
  }
}

/** Check multiple DOIs, preserving order; duplicates are only fetched once. */
export async function checkDois(dois: string[], fetch?: Fetcher): Promise<RetractionStatus[]> {
  const fetcher = fetch ?? buildFetcher();
  const cache = new Map<string, Promise<RetractionStatus>>();
  for (const doi of dois) {
    if (!cache.has(doi)) cache.set(doi, checkDoi(doi, fetcher));
  }
  const results = await Promise.all(dois.map((doi) => cache.get(doi)!));
  return results;
}

File: src/citeguard/checker.py
"""High-level API: check one or many DOIs, with optional caching."""

from __future__ import annotations

from .analyze import analyze_work
from .cache import FileCache
from .crossref import CrossrefError, DoiNotFoundError, Fetcher, build_fetcher
from .models import RetractionStatus, Verdict


def check_doi(doi: str, *, fetch: Fetcher | None = None, cache: FileCache | None = None) -> RetractionStatus:
    """Check a single DOI against Crossref and return its retraction status.

    Never raises for an expected failure (unknown DOI, network error) --
    those become a RetractionStatus with verdict NOT_FOUND or ERROR, so
    callers checking a whole bibliography can keep going past one bad
    entry.
    """
    fetch = fetch or build_fetcher()

    cached = cache.get(doi) if cache else None
    if cached is not None:
        return analyze_work(cached)

    try:
        message = fetch(doi)
    except DoiNotFoundError:
        return RetractionStatus(doi=doi, verdict=Verdict.NOT_FOUND, title=None, error="DOI not found in Crossref")
    except CrossrefError as exc:
        return RetractionStatus(doi=doi, verdict=Verdict.ERROR, title=None, error=str(exc))

    if cache:
        cache.set(doi, message)

    return analyze_work(message)


def check_dois(dois: list[str], *, fetch: Fetcher | None = None, cache: FileCache | None = None) -> list[RetractionStatus]:
    """Check multiple DOIs, preserving input order. Duplicate DOIs are only fetched once."""
    fetch = fetch or build_fetcher()
    results_by_doi: dict[str, RetractionStatus] = {}
    for doi in dois:
        if doi not in results_by_doi:
            results_by_doi[doi] = check_doi(doi, fetch=fetch, cache=cache)
    return [results_by_doi[doi] for doi in dois]
